File: 121-best-time-to-buy-and-sell-stock/best-time-to-buy-and-sell-stock.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int size = prices.size(); // Use .size() to get the number of elements
        if (size == 0) return 0;  // Edge case: If there are no prices, return 0

        int minPrice = prices[0]; // Initialize minPrice as the first day's price
        int maxProfit = 0;        // Initialize maxProfit to 0

        // Loop through the prices starting from the second day
        for (int i = 1; i < size; i++) {
            // Calculate profit if selling on the current day
            int profit = prices[i] - minPrice;

            // Update maxProfit if the current profit is higher
            if (profit > maxProfit) {
                maxProfit = profit;
            }

            // Update minPrice if the current price is lower
            if (prices[i] < minPrice) {
                minPrice = prices[i];
            }
        }

        return maxProfit;  // Return the maximum profit found
    }
};

File: 121-best-time-to-buy-and-sell-stock/best-time-to-buy-and-sell-stock.java
class Solution {
    public int maxProfit(int[] prices) {
        int n=prices.length;
        int cur_profit=0;
        int cur_price=prices[0];
        int max_profit=0;
        for(int i=1;i<n;i++){
            if(prices[i]<cur_price){
                cur_price=prices[i];
            }else{
                cur_profit=prices[i]-cur_price;
            }
            if(cur_profit>max_profit){
                max_profit=cur_profit;
            }
        }
        return max_profit;
    }
}